add intersection of two AABB_tree (#9501)
CGAL/cgal: AABB_tree/doc/AABB_tree/aabb_tree.txt

segments. An example of a distance query consists of finding the
closest point from a point query to a set of triangles.

This component solves the problem of finding all
intersecting pairs of objects and achieves better perfomance than
\ref chapterBoxIntersection "Intersecting Sequences of dD Iso-oriented Boxes"
when using parallel execution.

> 

Reviewer: ```suggestion
Similarly to the package \ref chapterBoxIntersection "Intersecting Sequences of dD Iso-oriented Boxes", this component might also be used for finding all
intersecting pairs of objects, and can even achieves better performances when running on several threads.

```

segments. An example of a distance query consists of finding the
closest point from a point query to a set of triangles.

> Similarly to the package \ref chapterBoxIntersection "Intersecting Sequences of dD Iso-oriented Boxes",

Reviewer 2: I am not sure, but does  the ref command not automatically put the text here?

Reviewer: Should be PkgBoxIntersectionD. Will push the fix.

this component might also be used for finding all intersecting pairs of objects,

> and can even achieves better performances when running on several threads.

Reviewer 2: ```suggestion
and can even achieve a better performance when running on several threads.
```

having degenerate triangles or segments in the AABB-tree will result in an undefined
behavior or a crash.

\section two_aabb_tree_interface Operations between Two AABB Trees

Given two AABB trees, the function `CGAL::AABB_trees::do_intersect()` determines whether any
primitive of the first tree intersects a primitive of the second tree.
To enumerate all intersecting pairs of primitives, use `CGAL::AABB_trees::all_pairs_of_intersecting_primitives()`.

These functions provide an alternative to `CGAL::box_intersection_d()`. In practice, `CGAL::AABB_trees::all_pairs_of_intersecting_primitives()` generally provides better performance,
particularly when using a parallel execution tag (see \ref aabb_tree_perf).

Note that compared to `CGAL::box_intersection_d()`, the tree of primitives is actually constructed. It becomes even more interesting in term of runtime

> if several calls to the aforementioned intersection functions are called with the same tree.

Reviewer: ```suggestion
if several calls to the aforementioned intersection functions are done with the same tree.
```

Reviewer 2: in term ->   in terms

having degenerate triangles or segments in the AABB-tree will result in an undefined
behavior or a crash.

\section two_aabb_tree_interface Operations between Two AABB Trees

Given two AABB trees, the function `CGAL::AABB_trees::do_intersect()` determines whether any
primitive of the first tree intersects a primitive of the second tree.
To enumerate all intersecting pairs of primitives, use `CGAL::AABB_trees::all_pairs_of_intersecting_primitives()`.

These functions provide an alternative to `CGAL::box_intersection_d()`. In practice, `CGAL::AABB_trees::all_pairs_of_intersecting_primitives()` generally provides better performance,
particularly when using a parallel execution tag (see \ref aabb_tree_perf).

Moreover, if several calls to `CGAL::AABB_trees::all_pairs_of_intersecting_primitives()` are performed using the same trees, the trees need to be constructed only once.

> Consequently, the performance advantage over `CGAL::box_intersection_d()` becomes even greater.

Reviewer: ```suggestion
Note that compared to `CGAL::box_intersection_d()`, the tree of primitives is actually constructed. It becomes even more interesting in term of runtime if several calls to the aforementioned intersection functions are called with the same tree.
```

triangles are stored into a single array as to form a triangle
soup. The primitive internally uses a `boost::iterator_adaptor`
as to provide the three functions `AABBPrimitive::id()`, `AABBPrimitive::datum()`,
and `AABBPrimitive::reference_point()` required by the primitive concept. In
\ref AABB_tree/AABB_custom_indexed_triangle_set_example.cpp "AABB_custom_indexed_triangle_set_example.cpp" the input is an indexed
triangle set stored through two arrays: one array of points and one
array of indices which refer to the point array. Here also the
primitive internally uses a `boost::iterator_adaptor`.

\subsection aabb_tree_examples_9 Intersection of Two Trees

In the following example, we compute all the intersections between two tetrahedra. The tetrahedra

> are stored as a vector of `Triangle_3`. We first compute if the tetrahedra do intersect and then we

Reviewer 2: So four of such triangles form one tetrahedron ????

Author: yes, is it not clear?

Reviewer 2: Well, we have a class `Tetahedron_3`.

Author: I think it is better to use something more generic than `Tetrahedron_3` in an example.

Reviewer: maybe say two triangle soups/meshes both representing a tetrahedron (if that's Andreas' point)

compute the indices of the intersecting triangles.

\cgalExample{AABB_tree/AABB_intersection_example.cpp}

\section aabb_tree_perf Performances

We provide some performance numbers for the case where the AABB tree
contains a set of polyhedron triangle facets. We measure the tree
construction time, the memory occupancy and the number of queries per
second for a variety of intersection and distance queries. The machine
used is a PC running Linux Mint with an Intel CPU Core Ultra 7 155H with 16 cores and 32GB of RAM.
By default, the kernel used is `Simple_cartesian<double>` (the fastest in our experiments). The

> program has been compiled with gcc 13.3 2023 compiler with the O3

Reviewer 2: ```suggestion
program has been compiled with the compiler gcc 13.3 2023 with the `O3`
```

option which maximizes speed.

\subsection aabb_tree_perf_cons Construction

The surface triangle mesh chosen for benchmarking the tree
construction is the knot model (14,400 triangles) depicted by
\cgalFigureRef{figAABB-tree-bench}. We measure the tree construction time (both

> AABB tree alone and AABB tree with internal KD-tree) for four denser versions if the knot model.

Reviewer 2: ```suggestion
AABB tree alone and AABB tree with internal KD-tree) for four denser versions of the knot model.
```

`AABB_tree::closest_point_and_primitive()`) from point queries randomly chosen
inside the bounding box.

|  Nb triangles | Closest_point() | Squared_distance() | Closest_point_and_primitive() |
| ----:    | ----: | ----: | -: |
|   14,400 | 196,043 | 196,356 | 196,708 |
|   57,600 | 123,388 | 123,844 | 123,567 |
|  230,400 | 54,480 | 53,235 | 54,329 |
|  921,600 | 25,761 | 26,890 | 25,386 |
|  3,686,400 | 13,696 | 13,389 | 13,440 |
|  14,745,600 | 7,158 | 6,913| 7,010 |

\subsection aabb_two_trees_perf Intersection of Two AABB Trees

> We test the intersection of two AABB trees using 3 models: iphigenia and the Nefertiti that are two common models of mesh processing and twisted knot to maximize the number of intersection. For each model,

Reviewer: ```suggestion
We test the intersection of two AABB trees using 3 models: two common models of mesh processing (Iphigenia and Nefertiti), and twisted knot to maximize the number of intersection. For each model,
```

we test their intersection with a translated version of themselves (Note: We do not exploit the fact that it is the same model, the tree is built twice).
We compare here `CGAL::AABB_trees::all_pairs_of_intersecting_primitives()` and `CGAL::box_intersection_d()`.

> We exhibit the performance of the intersection of two AABB trees using three models: Iphigenia and Nefertiti \cgalFigureRef{figAABB-tree-bench}, which are widely used benchmark meshes in geometry processing, and a twisted knot \cgalFigureRef{figAABB-tree-bench}, chosen to maximize the number of intersections.

Reviewer: ```suggestion
```

> For each model, we compute the intersections between the model and a translated copy of itself. Although both trees are built from the same mesh, we intentionally construct the AABB tree twice and do not exploit this fact.

Reviewer: ```suggestion
```

> We compare the performance of `CGAL::AABB_trees::all_pairs_of_intersecting_primitives()` with that of `CGAL::box_intersection_d()`.

Reviewer: ```suggestion
```

`AABB_tree::closest_point_and_primitive()`) from point queries randomly chosen
inside the bounding box.

|  Nb triangles | Closest_point() | Squared_distance() | Closest_point_and_primitive() |
| ----:    | ----: | ----: | -: |
|   14,400 | 196,043 | 196,356 | 196,708 |
|   57,600 | 123,388 | 123,844 | 123,567 |
|  230,400 | 54,480 | 53,235 | 54,329 |
|  921,600 | 25,761 | 26,890 | 25,386 |
|  3,686,400 | 13,696 | 13,389 | 13,440 |
|  14,745,600 | 7,158 | 6,913| 7,010 |

\subsection aabb_two_trees_perf Intersection of Two AABB Trees

We test the intersection of two AABB trees using 3 models: iphigenia and the Nefertiti that are two common models of mesh processing and twisted knot to maximize the number of intersection. For each model,

> we test their intersection with a translated version of themselves (Note: We do not exploit the fact that it is the same model, the tree is built two times).

Reviewer: ```suggestion
we test their intersection with a translated version of themselves (Note: We do not exploit the fact that it is the same model, the tree is built twice).
```
